### PR TITLE
feat(#189): remove instance number on doNames for pDO in ExtRef while…

### PR DESCRIPTION
**Pull request by AliouDIAITE**
Signed-off-by: Aliou DIAITE <aliou.diaite@rte-france.com>

### CODE REVIEW

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java [hunk 1]
```diff
@@ -239,13 +241,119 @@ private static TPrivate createPrivate(JAXBElement<?> jaxbElement) {
...
+     *
+     * @param stds list of STD to short
+     * @return map of ICD_SYSTEM_VERSION_UUID attribute in IED/Private:COMPAS-ICDHeader and related Private coupled with
+     * all corresponding STD
+     */
+    public static Map<String, Pair<TPrivate, List<SCL>>> createMapICDSystemVersionUuidAndSTDFile(Set<SCL> stds) {
+        Map<String, Pair<TPrivate, List<SCL>>> stringSCLMap = new HashMap<>();
+        stds.forEach(std -> std.getIED().forEach(ied -> ied.getPrivate().forEach(tp ->
+                PrivateService.extractCompasICDHeader(tp).map(TCompasICDHeader::getICDSystemVersionUUID).ifPresent(icdSysVer -> {
+                    Pair<TPrivate, List<SCL>> pair = stringSCLMap.get(icdSysVer);
```
[the remark this answers is not in the document]
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java:255]: ok

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java [hunk 2]
```diff
@@ -239,13 +241,119 @@ private static TPrivate createPrivate(JAXBElement<?> jaxbElement) {
...
 
+    /**
+     * Sorts in map of ICD_SYSTEM_VERSION_UUID and related Private coupled with all corresponding STD for all given STD
+     *
+     * @param stds list of STD to short
+     * @return map of ICD_SYSTEM_VERSION_UUID attribute in IED/Private:COMPAS-ICDHeader and related Private coupled with
+     * all corresponding STD
+     */
+    public static Map<String, Pair<TPrivate, List<SCL>>> createMapICDSystemVersionUuidAndSTDFile(Set<SCL> stds) {
+        Map<String, Pair<TPrivate, List<SCL>>> stringSCLMap = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java:252]: ok

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/SclService.java [hunk 3]
```diff
@@ -499,10 +497,10 @@ public static Set<Pair<Integer, String>> getEnumTypeElements(SCL scd, String idE
     public static SclRootAdapter importSTDElementsInSCD(@NonNull SclRootAdapter scdRootAdapter, Set<SCL> stds, Map<Pair<String, String>, List<String>> comMap) throws ScdException {
 
         //Check SCD and STD compatibilities
-        Map<String, Pair<TPrivate, List<SCL>>> mapICDSystemVersionUuidAndSTDFile = createMapICDSystemVersionUuidAndSTDFile(stds);
-        checkSTDCorrespondanceWithLNodeCompasICDHeader(mapICDSystemVersionUuidAndSTDFile);
+        Map<String, PrivateService.PrivateLinkedToSTDs> mapICDSystemVersionUuidAndSTDFile = PrivateService.createMapICDSystemVersionUuidAndSTDFile(stds);
+        PrivateService.checkSTDCorrespondanceWithLNodeCompasICDHeader(mapICDSystemVersionUuidAndSTDFile);
```
Comment (gleizesDor) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/SclService.java:501]:
```suggestion
        List<PrivateService.PrivateLinkedToSTDs> pToStds = PrivateService.createMapICDSystemVersionUuidAndSTDFile(stds).toList();
        pToStds.stream()
                .filter(pToSTD -> pToSTD.stdList().size() != 1)
                .findFirst()
                .ifPresent(pToStd -> {
                    throw new ScdException("There are several STD files corresponding to " + stdCheckFormatExceptionMessage(pToStd.tPrivate()));
                });
```

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/SclService.java [hunk 4]
```diff
@@ -512,14 +510,14 @@ public static SclRootAdapter importSTDElementsInSCD(@NonNull SclRootAdapter scdR
             );
 
             if (!mapICDSystemVersionUuidAndSTDFile.containsKey(icdSysVerUuid))
```
Comment (gleizesDor) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/SclService.java:512]:
```suggestion
PrivateService.PrivateLinkedToSTDs pToStd = pToStds
                    .stream()
                    .filter(pTStd -> PrivateService.extractCompasICDHeader(pTStd.tPrivate()).map(TCompasICDHeader::getICDSystemVersionUUID).orElseThrow().equals(icdSysVerUuid))
                    .findFirst()
                    .orElseThrow(() -> new ScdException("There is no STD file found corresponding to " + stdCheckFormatExceptionMessage(tPrivate)));
            // import /ied /dtt in Scd
            SCL std = pToStd.stdList().get(0);
```

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java [hunk 5]
```diff
@@ -239,13 +241,122 @@ private static TPrivate createPrivate(JAXBElement<?> jaxbElement) {
...
+    /**
+     * Checks SCD and STD compatibilities by checking if there is at least one ICD_SYSTEM_VERSION_UUID in
+     * Substation/../LNode/Private COMPAS-ICDHeader of SCL not present in IED/Private COMPAS-ICDHeader of STD
+     *
+     * @param mapICDSystemVersionUuidAndSTDFile map of ICD_SYSTEM_VERSION_UUID and list of corresponding STD
+     * @throws ScdException throws when there are several STD files corresponding to <em>ICD_SYSTEM_VERSION_UUID</em>
+     *                      from Substation/../LNode/Private COMPAS-ICDHeader of SCL
+     */
+    public static void checkSTDCorrespondanceWithLNodeCompasICDHeader(Map<String, PrivateLinkedToSTDs> mapICDSystemVersionUuidAndSTDFile) throws ScdException {
+        for (PrivateLinkedToSTDs privateLinkedToSTDs : mapICDSystemVersionUuidAndSTDFile.values()) {
```
Comment (gleizesDor) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java:278]: use stream + filter + findFirst

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java [hunk 6]
```diff
@@ -239,13 +241,122 @@ private static TPrivate createPrivate(JAXBElement<?> jaxbElement) {
...
+    }
 
 
+    /**
+     * Creates map of IEDName and related Private for all Privates COMPAS-ICDHeader in /Substation of SCL
+     *
+     * @param scdRootAdapter SCL file in which Private should be found
+     * @return map of Private and its IEDName parameter
+     */
+    public static Map<String, TPrivate> createMapIEDNameAndPrivate(SclRootAdapter scdRootAdapter) {
```
Comment (gleizesDor) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java:308]: this method is only used to populate the `scdRootAdapter` so we could return a stream of `Stream<TPrivate>` and use a `foreach` on this stream

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/SclService.java (hunk 3 above)
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/SclService.java:501]: Not work

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java [hunk 7]
```diff
@@ -239,13 +241,122 @@ private static TPrivate createPrivate(JAXBElement<?> jaxbElement) {
...
+                    list.add(std);
+                    stringSCLMap.put(icdSysVer, new PrivateLinkedToSTDs(tp, list));
+                })
+        )));
+        return stringSCLMap;
+    }
 
 
+    public record PrivateLinkedToSTDs (TPrivate tPrivate, List<SCL> stdList) {
+    }
```
[the remark this answers is not in the document]
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java:266]: Not work

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/SclService.java [hunk 8]
```diff
@@ -499,10 +497,10 @@ public static Set<Pair<Integer, String>> getEnumTypeElements(SCL scd, String idE
     public static SclRootAdapter importSTDElementsInSCD(@NonNull SclRootAdapter scdRootAdapter, Set<SCL> stds, Map<Pair<String, String>, List<String>> comMap) throws ScdException {
 
         //Check SCD and STD compatibilities
-        Map<String, Pair<TPrivate, List<SCL>>> mapICDSystemVersionUuidAndSTDFile = createMapICDSystemVersionUuidAndSTDFile(stds);
-        checkSTDCorrespondanceWithLNodeCompasICDHeader(mapICDSystemVersionUuidAndSTDFile);
+        Map<String, PrivateService.PrivateLinkedToSTDs> mapICDSystemVersionUuidAndSTDFile = PrivateService.createMapICDSystemVersionUuidAndSTDFile(stds);
```
[the remark this answers is not in the document]
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/SclService.java:500]: The work should maybe done for another object, but i choosed a map to easily regroup elements.
No! we need to know occurrence of private with icdSystemVersionUuid and linked STD file in which we find them. (Your comment show that message and logic should be rechecked)

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/SclService.java (hunk 4 above)
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/SclService.java:512]: Linked to change suggestion you made, so application is to be rechecked

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java (hunk 6 above)
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java:308]: ok

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java (hunk 5 above)
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java:278]: ok

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java [hunk 9]
```diff
@@ -239,13 +241,122 @@ private static TPrivate createPrivate(JAXBElement<?> jaxbElement) {
...
 
 
+    /**
+     * Sorts in map of ICD_SYSTEM_VERSION_UUID and related Private coupled with all corresponding STD for all given STD
+     *
+     * @param stds list of STD to short
+     * @return map of ICD_SYSTEM_VERSION_UUID attribute in IED/Private:COMPAS-ICDHeader and related Private coupled with
+     * all corresponding STD
+     */
+    public static Map<String, PrivateLinkedToSTDs> createMapICDSystemVersionUuidAndSTDFile(Set<SCL> stds) {
```
[the remark this answers is not in the document]
Comment (AliouDIAITE) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/PrivateService.java:250]: Refactoring will be done in issue #110

**Review by massifben — APPROVED** [review]

**Review by samirromdhani — APPROVED** [review]

**Pull request merged**